File: JavaGames/src/main/java/com/josiahcampbell/algos/BinarySearch.java
package com.josiahcampbell.algos;

import java.util.List;

public class BinarySearch<T extends Comparable<T>> {

    public int indexOf(List<T> elements, T searchElement) {
        int low = 0;
        int high = elements.size() - 1;
        while (low <= high) {
            int mid = (low + high) / 2;
            if (elements.get(mid).equals(searchElement)) {
                return mid;
            } else if (elements.get(mid).compareTo(searchElement) > 0) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return -1;
    }


    public int search(int[] elements, int element) {
        int low = 0;
        int high = elements.length - 1;
        while (low <= high) {
            int mid = (low + high) / 2;
            if (elements[mid] == element) {
                return mid;
            } else if (elements[mid] < element) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return -1;
    }
}

File: ruby_games/lib/algos/binary_search.rb
class BinarySearch
  def index_of(numbers, item)
    low = 0
    high = numbers.length - 1
    while low <= high do
      mid = (low + high) / 2
      if numbers[mid] == item
        return mid
      elsif numbers[mid] < item
        low += 1
      else
        high -= 1
      end
    end
    return -1
  end
end